x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java

```diff
@@ -143,6 +160,7 @@ protected void done(boolean success) {
         synchronized (this) {
             assert completed == false;
             completed = true;
+            assert listeners == null || listeners.stream().allMatch(l -> progress < l.position() && l.position() <= end);
```

Comment:
Relates https://github.com/elastic/elasticsearch-serverless/pull/1872#pullrequestreview-2033093453

Reply:
Hmm the actual length of the data is `actionResult()` which should be <= `end`. When `actionResult() < end`, I think there should be no listener waiting at all when we reach here? Otherwise these listeners are waiting for a range that can never be fulfilled?

Reply:
> the actual length of the data is `actionResult()` which should be <= `end`.

I think that the cache does not allow to partially complete ranges, so it expects the range (gap) to be fully written or it considers it failed and any remaining listener will be failed too.

I pushed cdabee72d52b3c814c72f71cbd4296c5b2e329db to ensure it is always completed with a value equals to `end`.